server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -400,7 +400,16 @@ private static BulkItemResponse processUpdateResponse(final UpdateRequest update
                 if (response.getFailure().getSeqNo() == SequenceNumbers.UNASSIGNED_SEQ_NO) {
                     continue; // ignore replication as we didn't generate a sequence number for this request.
                 }
-                operationResult = replica.markSeqNoAsNoop(response.getFailure().getSeqNo(), response.getFailure().getMessage());
+
+                final long primaryTerm;
+                if (response.getFailure().getTerm() == SequenceNumbers.UNASSIGNED_PRIMARY_TERM) {
+                    // primary is on older version, just take the current primary term
+                    primaryTerm = replica.getOperationPrimaryTerm();
```

Comment:
I think we know the correct term to use in the `ConcreteReplicaRequest`, unfortunately we lost that info here. I wonder if it could be worth exposing the primary term to the subclasses of `TransportReplicationAction`? We could also then assert that the bulk operations primaryTerm is the same as all the result primary terms?

Could be done in a follow-up if you agree, no need to include it into this PR.

Reply:
> I wonder if it could be worth exposing the primary term to the subclasses of TransportReplicationAction? We could also then assert that the bulk operations primaryTerm is the same as all the result primary terms?

This is currently enforced by assertions in `IndexShard` which check that the operationPrimaryTerm is lower or equal to the current term of the shard (which is determined by the term in `ConcreteReplicaRequest`).

This lower or equal is a sufficient condition, and depending on where we go with the dirty reads work, we want these two terms not to be always equal, but op-primary-term truly lower in some cases in the future.